File: 0744-network-delay-time/0744-network-delay-time.java
import java.util.Arrays;

class Solution {
    public int networkDelayTime(int[][] times, int n, int k) {
        int[] dis = new int[n];
        Arrays.fill(dis, Integer.MAX_VALUE);
        dis[k - 1] = 0;  

        for (int i = 0; i < n - 1; i++) {
            for (int[] time : times) {
                int u = time[0] - 1;  
                int v = time[1] - 1;  
                int w = time[2];   

                if (dis[u]!= Integer.MAX_VALUE && dis[u] + w < dis[v]) {
                    dis[v] = dis[u] + w;
                }
            }
        }

        
        int res = -1;
        for (int i : dis) {
            if (i == Integer.MAX_VALUE) {
                return -1;  
            }
            System.out.println(i);
            res = Math.max(res, i); 
        }

        return res;
    }
}

File: 0744-network-delay-time/0744-network-delay-time.py
from heapq import heappush, heappop
from collections import defaultdict

class Solution(object):
    def networkDelayTime(self, times, n, k):
        """
        :type times: List[List[int]]
        :type n: int
        :type k: int
        :rtype: int
        """
        adj = defaultdict(list)
        for u, v, w in times:
            adj[u].append((v, w))

        heap = [(0, k)]  # (distance, node)
        visited = set()
        res = 0

        while heap:
            curr_dis, curr_node = heappop(heap)
            if curr_node in visited:
                continue
            visited.add(curr_node)
            res = max(res, curr_dis)

            for neighbor, weight in adj[curr_node]:
                if neighbor not in visited:
                    heappush(heap, (curr_dis + weight, neighbor))

        return res if len(visited) == n else -1
